File: day9/three_largest_numbers.java
/*
  Write a function that takes in an array of at least three integers and,
  without sorting the input array, returns a sorted array of the three largest
  integers in the input array.
 */


/* Time Complexity: O(n)
Space Complexity: O(1)
 */

import java.util.*;

class Program {
  public static int[] findThreeLargestNumbers(int[] array) {
    int[] largest_nums = {Integer.MIN_VALUE, Integer.MIN_VALUE, Integer.MIN_VALUE};
    for(int i = 0; i< array.length; i++) {
      if(largest_nums[2] == Integer.MIN_VALUE || array[i] > largest_nums[2]) {
        updateLargestNums(largest_nums, array[i], 2);
      }
      else if(largest_nums[1] == Integer.MIN_VALUE || array[i] > largest_nums[1]) {
        updateLargestNums(largest_nums, array[i], 1);
      }
       else if(largest_nums[0] == Integer.MIN_VALUE || array[i] > largest_nums[0]) {
        updateLargestNums(largest_nums, array[i], 0);
      }
    }
    return largest_nums;
  }

  public static void updateLargestNums(int[] array, int num, int position) {
    int index = 0;
    while(index <= position) {
      if(index == position) {
        array[position] = num;
      }
      else{
        array[index] = array[index + 1];
      }
      index++;
    }
  }
}

File: day9/three_largest_numbers.py

#   Write a function that takes in an array of at least three integers and,
#   without sorting the input array, returns a sorted array of the three largest
#   integers in the input array.


# Approach 1
# Time Complexity: O(n)
# Space Complexity: O(1)
def findThreeLargestNumbers(array):
    largest_nums = [None] * 3
    for i in reversed(range(len(largest_nums))):
        max_num = max(array)
        largest_nums[i] = max_num
        array.remove(max_num)
    return largest_nums


# Approach 2: Better One
# Time Complexity: O(n)
# Space Complexity: O(1)
def findThreeLargestNumbers(array):
    largest_nums = [None]*3
    for num in array:
        if largest_nums[2] is None or largest_nums[2] < num:
            updateLargestNums(largest_nums, num, 2)
        elif largest_nums[1] is None or largest_nums[1] < num:
            updateLargestNums(largest_nums, num, 1)
        elif largest_nums[0] is None or largest_nums[0] < num:
            updateLargestNums(largest_nums, num, 0)
    return largest_nums


def updateLargestNums(largest_nums, num, position):
    index = 0
    while (index <= position):
        if index == position:
            largest_nums[position] = num
        else:
            largest_nums[index] = largest_nums[index + 1]
        index += 1
